### ISSUE TITLE
Services in containers not running

### ISSUE THREAD

**Issue by santosh0705** [adelton/docker-freeipa#66]
I'm getting strange behavior on Debian 8. When the system restart network services in the containers are not biding to any ports. Sometimes it works fine (Takes too long time to startup) but most of the time its not.

Below is the output of netstat command:

```
[root@ipa /]# netstat -ntple
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name
tcp6       0      0 127.0.0.1:8005          :::*                    LISTEN      17         15141      -
tcp6       0      0 127.0.0.1:8009          :::*                    LISTEN      17         11841      -
tcp6       0      0 :::8080                 :::*                    LISTEN      17         14622      -
tcp6       0      0 :::8443                 :::*                    LISTEN      17         11840      -
```

When starting/restarting manually it works fine. Seems like it depends on any system services which is not available during startup.

**Comment by adelton** [comment]
What does
```
systemctl
```
and
```
systemctl status
```
in the container say?

**Comment by santosh0705** [comment]
Output of `systemctl`:
```
  UNIT                                      LOAD   ACTIVE     SUB     JOB   DESCRIPTION
  -.mount                                   loaded active     mounted       /
  data.mount                                loaded active     mounted       /data
  dev-mqueue.mount                          loaded active     mounted       POSIX Message Queue File System
  etc-hostname.mount                        loaded active     mounted       /etc/hostname
  etc-hosts.mount                           loaded active     mounted       /etc/hosts
  etc-resolv.conf.mount                     loaded active     mounted       /etc/resolv.conf
  proc-bus.mount                            loaded active     mounted       /proc/bus
  proc-fs.mount                             loaded active     mounted       /proc/fs
  proc-irq.mount                            loaded active     mounted       /proc/irq
  proc-kcore.mount                          loaded active     mounted       /proc/kcore
  proc-sysrq\x2dtrigger.mount               loaded active     mounted       /proc/sysrq-trigger
  proc-timer_stats.mount                    loaded active     mounted       /proc/timer_stats
  tmp.mount                                 loaded active     mounted       Temporary Directory
  usr-share-zoneinfo-Etc-UTC.mount          loaded active     mounted       /usr/share/zoneinfo/Etc/UTC
  certmonger.service                        loaded active     running       Certificate monitoring and PKI enrollment
  dbus.service                              loaded active     running       D-Bus System Message Bus
* dirsrv@LOCUZ-COM.service                  loaded failed     failed        389 Directory Server LOCUZ-COM.
  ipa-server-update-self-ip-address.service loaded activating start   start Update self IP address of the IPA server
* ipa.service                               loaded failed     failed        Identity, Policy, Audit
* ntpd.service                              loaded failed     failed        Network Time Service
  pki-tomcatd@pki-tomcat.service            loaded active     running       PKI Tomcat Server pki-tomcat
  sssd.service                              loaded active     running       System Security Services Daemon
  systemd-journald.service                  loaded active     running       Journal Service
  systemd-tmpfiles-setup.service            loaded active     exited        Create Volatile Files and Directories
  -.slice                                   loaded active     active        Root Slice
  system-dirsrv.slice                       loaded active     active        system-dirsrv.slice
  system-pki\x2dtomcatd.slice               loaded active     active        system-pki\x2dtomcatd.slice
  system.slice                              loaded active     active        System Slice
  dbus.socket                               loaded active     running       D-Bus System Message Bus Socket
  systemd-journald-dev-log.socket           loaded active     running       Journal Socket (/dev/log)
  systemd-journald.socket                   loaded active     running       Journal Socket
  container-ipa.target                      loaded active     active        Minimal target for containerized FreeIPA server
  dirsrv.target                             loaded active     active        389 Directory Server
  network.target                            loaded active     active        Network
  nss-user-lookup.target                    loaded active     active        User and Group Name Lookups
  pki-tomcatd.target                        loaded active     active        PKI Tomcat Server

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.
JOB    = Pending job for the unit.

36 loaded units listed. Pass --all to see loaded but inactive units, too.
To show all installed unit files use 'systemctl list-unit-files'.
```

Output of `systemctl status`:

```
* ipa.locuz.com
    State: initializing
     Jobs: 1 queued
   Failed: 3 units
    Since: Mon 2016-04-04 09:30:23 IST; 9min ago
   CGroup: /docker/6b05a761a82f247de9a5825854b0e21b5e5a38cf8c622d1f402139f90d91e306
           |-  1 /usr/sbin/init --show-status=false
           |-491 bash
           |-528 systemctl status
           `-system.slice
             |-dbus.service
             | `-69 /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation
             |-system-pki\x2dtomcatd.slice
             | `-pki-tomcatd@pki-tomcat.service
             |   `-272 /usr/lib/jvm/jre/bin/java -DRESTEASY_LIB=/usr/share/java/resteasy -Djava.library.path=/usr/lib64/nuxwdog-jni -classpath /usr/share/tomcat/bin/bootstrap.jar:/usr/share/tomcat/bin/tomcat-juli.jar:/usr/lib/java/commons-daemon.jar -Dcatalina.base=/var/lib/pki/pki-tomcat -Dcatalina.home=/usr/share/tomcat -Djava.endorsed.dirs= -Djava.io.tmpdir=/var/lib/pki/pki-tomcat/temp -Djava.util.logging.config.file=/var/lib/pki/pki-tomcat/conf/logging.properties -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager -Djava.security.manager -Djava.security.policy==/var/lib/pki/pki-tomcat/conf/catalina.policy org.apache.catalina.startup.Bootstrap start
             |-systemd-journald.service
             | `-68 /usr/lib/systemd/systemd-journald
             |-sssd.service
             | |-176 /usr/sbin/sssd -D -f
             | |-282 /usr/libexec/sssd/sssd_be --domain locuz.com --uid 0 --gid 0 --debug-to-files
             | |-296 /usr/libexec/sssd/sssd_nss --uid 0 --gid 0 --debug-to-files
             | |-297 /usr/libexec/sssd/sssd_sudo --uid 0 --gid 0 --debug-to-files
             | |-298 /usr/libexec/sssd/sssd_pam --uid 0 --gid 0 --debug-to-files
             | `-299 /usr/libexec/sssd/sssd_pac --uid 0 --gid 0 --debug-to-files
             |-ipa-server-update-self-ip-address.service
             | |-407 /bin/bash /usr/sbin/ipa-server-configure-first update-self-ip-address
             | |-410 tail --silent -n 0 -f --retry /var/log/ipa-server-configure-first.log
             | `-524 host -t A ipa.locuz.com
             `-certmonger.service
               `-77 /usr/sbin/certmonger -S -p /var/run/certmonger.pid -n
```

Output of `netstat -ntple`:

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name    
tcp6       0      0 127.0.0.1:8005          :::*                    LISTEN      17         16776      -                   
tcp6       0      0 127.0.0.1:8009          :::*                    LISTEN      17         15536      -                   
tcp6       0      0 :::8080                 :::*                    LISTEN      17         10834      -                   
tcp6       0      0 :::8443                 :::*                    LISTEN      17         15535      -
```

**Comment by adelton** [comment]
It looks like the problems stem from the dirsrv not having started properly. You might want to check its status and logs to see what is the problem. Basically, whatever debugging / investigation you'd do on FreeIPA server outside of container would apply here as well.

**Comment by santosh0705** [comment]
I end up with `systemd` service startup script. Adding 10 seconds delay before container start. 5 seconds didn't work :disappointed: 

Below is the example file `/lib/systemd/system/freeipa.service`:

```
[Unit]
Description=FreeIPA server container
Requires=docker.service
After=docker.service

[Service]
ExecStartPre=/bin/sh -c '/usr/bin/test -f /var/run/freeipa-container || /bin/sleep 10'
ExecStartPost=/usr/bin/touch /var/run/freeipa-container
ExecStart=/usr/bin/docker start -a ipa-server
ExecStop=/usr/bin/docker stop -t 5 ipa-server

[Install]
WantedBy=basic.target
```

Could not found any better solution to fix this problem. Let me know if someone get a better solution to this issue.

**Comment by adelton** [comment]
I wonder if you might want to go with some dependency on some other service which is still just starting at that point, rather than a hardcoded `sleep`. But I'm glad that you've found solution that works for you.

**Comment by santosh0705** [comment]
I checked with docker and network services but can't figure out it depends on which service. Even putting 5 seconds delay it didn't work. I assume in 5 seconds all the services were up and running.

I'm not sure but doubt it is `/dev/random` related issue in container, may be some service in container using is for cryptography and due to its delayed response nature in container services may be failing.

On free time I'll try to re-generate this issue in different system and get back again if something I find.

**Comment by adelton** [comment]
OK. Closing now, feel free to reopen or file new ticket if you feel there is something that ought to be done in the docker-freeipa container image.

**Comment by santosh0705** [comment]
Even after a delay of 10 secs it got failed yesterday on system reboot.

The services `dirsrv@XXXX-COM.service`, `ipa.service` & `ntpd.service` didn't start like before and `ipa-server-update-self-ip-address.service` in activating stat.

`dirsrv` didn't start and its log saying:
```
[05/May/2016:19:29:43 +051800] - Unable to start slapd because it is already running as process 306
[05/May/2016:19:29:43 +051800] - Shutting down due to possible conflicts with other slapd processes
```

`ntpd` is failing because name resolution is not happening since the `dirsrv` is not started.

Restarting the `dirsrv` service worked only after deleting the lock file:
`/var/run/lock/dirsrv/slapd-LOCUZ-COM/server/306`

In `journalctl` I see `dirsrv` restarted in 16 sec time interval, may be the 306 is the pid of the previously started service.

No idea why it is happening 😞

**Comment by ToBeReplaced** [comment]
I'm hitting this issue. On reboot, *sometimes* the dirsrv fails to start, and then errors cascade. I see the same log messages as above; `Unable to start slapd because it is already running as process XXX`. The process does not actually exist, and removing `/var/run/lock/dirsrv/slapd-EXAMPLE-COM/server/XXX` followed by `systemctl restart ipa` alleviates the issue.

Maybe we should remove that file on start?

Also, I noticed the comment about `/dev/random`. For me, `kadmin` can fail to start if there is not enough entropy (I think it times out after 5 minutes...).

**Comment by adelton** [comment]
I believe 1fa29dc89b7ea8b364e47b558495d63ba6e01e5c which is in master/fedora-23 branches now address the issue. We had an exception for purging `/run/lock` in the code, I've removed it now.
